torch/csrc/distributed/c10d/Work.cpp

```diff
@@ -70,7 +71,10 @@ std::vector<at::Tensor> Work::result() {
   TORCH_CHECK(false, "result() not implemented.");
 }
 
-void Work::synchronize() {}
+void Work::synchronize() {
+  c10d::unregister_work(
+      c10::intrusive_ptr<Work>::unsafe_reclaim_from_nonowning(this));
+}
```

Comment:
seems both `wait()` and `synchronize()` should serve the same purpose (similar reasoning as in WorkNCCL)